src/commands/announce.ts

```diff
@@ -50,66 +54,49 @@ module.exports = new Command({
-					{ name: 'tank', arr: lobby.tankPlayers },
-					{ name: 'damage', arr: lobby.damagePlayers },
-					{ name: 'support', arr: lobby.supportPlayers },
-				].sort((a, b) => {
-					if (a.arr.length < b.arr.length) return -1
-					if (a.arr.length > b.arr.length) return 1
-
-					return 0
-				})
-
-				if (findSignup[rolePools[0].name + 'Rank'] === lobby.rank) {
-					lobby[rolePools[0].name + 'Players'].push(findSignup)
-					continue
-				}
-
-				if (findSignup[rolePools[1].name + 'Rank'] === lobby.rank) {
-					lobby[rolePools[1].name + 'Players'].push(findSignup)
-					continue
-				}
-
-				if (findSignup[rolePools[2].name + 'Rank'] === lobby.rank) {
-					lobby[rolePools[2].name + 'Players'].push(findSignup)
-					continue
-				}
+			if (findSignup && findSignup.confirmedOn) {
+				// Add user to role pool, if they have a role placed in the correct rank
+				if (findSignup.tankRank === lobby.rank) lobby.tankPlayers.push(findSignup)
+				if (findSignup.damageRank === lobby.rank) lobby.damagePlayers.push(findSignup)
+				if (findSignup.supportRank === lobby.rank) lobby.supportPlayers.push(findSignup)
 			}
 		}
 
-		lobby.tankPlayers.sort((a, b) => sortPlayers(a, b, lobby))
-		lobby.damagePlayers.sort((a, b) => sortPlayers(a, b, lobby))
-		lobby.supportPlayers.sort((a, b) => sortPlayers(a, b, lobby))
-
-		const topTanks = lobby.tankPlayers.slice(0, tankCount)
-		const topDps = lobby.damagePlayers.slice(0, dpsCount)
-		const topSupports = lobby.supportPlayers.slice(0, suppCount)
-
-		topTanks.forEach(s => {
-			guildMembers.get(s.discordId)?.roles.add(ingameRole)
-			mongoSignups.updateOne({ discordId: s.discordId }, { $inc: { gamesPlayed: 1 } })
-		})
-
-		topDps.forEach(s => {
-			guildMembers.get(s.discordId)?.roles.add(ingameRole)
-			mongoSignups.updateOne({ discordId: s.discordId }, { $inc: { gamesPlayed: 1 } })
-		})
-
-		topSupports.forEach(s => {
-			guildMembers.get(s.discordId)?.roles.add(ingameRole)
-			mongoSignups.updateOne({ discordId: s.discordId }, { $inc: { gamesPlayed: 1 } })
+		// Sort roles by which has the least players already
+		const rolePools = Object.fromEntries(
+			Object.entries({
+				tank: { arr: lobby.tankPlayers, possibleArr: <Signup[]>[], lockedArr: <Signup[]>[] },
+				damage: { arr: lobby.damagePlayers, possibleArr: <Signup[]>[], lockedArr: <Signup[]>[] },
+				support: { arr: lobby.supportPlayers, possibleArr: <Signup[]>[], lockedArr: <Signup[]>[] },
+			}).sort((a, b) => {
+				if (a[1].arr.length < b[1].arr.length) return -1
+				if (a[1].arr.length > b[1].arr.length) return 1
+
+				return 0
+			}),
+		)
+
+		Object.entries(rolePools).forEach(([name, rolePool]) => {
+			// Sort and filter those who aren't locked
+			rolePool.arr.sort((a, b) => sortPlayers(a, b, lobby))
+			rolePool.possibleArr = rolePool.arr.filter(signup => !signup.playing)
+			rolePool.lockedArr = rolePool.possibleArr.slice(0, counts[name as keyof typeof counts])
+			rolePool.lockedArr.forEach(signup => (signup.playing = true))
```

Comment:
This looks like wizardry, I can't follow this in my head :D

Reply:
See matchmaker.md, but yeah I should add some comments